### PR TITLE
Adding in ability to support a library for freertos_config and a custom port

### PR DESCRIPTION
**Pull request by phelter**
Description
-----------

**DEPRECATE:** CMake: `FREERTOS_CONFIG_FILE_DIRECTORY` configuration and replace it with expected library `freertos_config`
Rationale - some configuration settings can now be added to the `freertos_config` target eg:

```cmake
target_compile_definitions(freertos_config
  INTERFACE
    projCOVERAGE_TEST=0
)
```

Added support for `FREERTOS_PORT=A_CUSTOM_PORT` - this allows the inclusion of a port as a library as well that can be linked to the FreeRTOS-Kernel appropriately.

Assorted fixes:
* spelling: `Compiller` -> `Compiler`
* missing link to Threads::Threads (pthreads) for GCC_POSIX build.

Test Steps
-----------
Setup a local cmake project and use the following project file to include FreeRTOS-Kernel

```cmake
FetchContent_Declare( freertos_kernel
  GIT_REPOSITORY git@github.com:phelter/FreeRTOS-Kernel.git
  GIT_TAG        feature/cmake-updates
)

FetchContent_Declare( freertos_plus_tcp
  GIT_REPOSITORY git@github.com:phelter/FreeRTOS-Plus-TCP.git
  GIT_TAG        cmake-support
  GIT_SUBMODULES "" # Don't grab any submodules since not latest
)

add_library(freertos_config INTERFACE)
add_library(FreeRTOS::Config ALIAS freertos_config)

target_sources(freertos_config
  INTERFACE
    inc/FreeRTOSConfig.h
    inc/FreeRTOSIPConfig.h
)

target_include_directories(freertos_config SYSTEM
INTERFACE
    inc
)

target_compile_definitions(freertos_config
  INTERFACE
    projCOVERAGE_TEST=0
)
# -----------------------------------------------------------------------------
set( FREERTOS_HEAP "4" CACHE STRING "" FORCE)
set( FREERTOS_PORT "GCC_POSIX" CACHE STRING "" FORCE)
if (CMAKE_CROSSCOMPILING)
  set(FREERTOS_PORT "GCC_ARM_CA9" CACHE STRING "" FORCE)
endif()

FetchContent_MakeAvailable(freertos_kernel)

# -----------------------------------------------------------------------------
set(FREERTOS_PLUS_TCP_BUFFER_ALLOCATION "1" CACHE STRING "" FORCE)
set(FREERTOS_PLUS_TCP_COMPILER "" CACHE STRING "" FORCE)
set( FREERTOS_PLUS_TCP_NETWORK_IF "POSIX" CACHE STRING "" FORCE)
if (CMAKE_CROSSCOMPILING)
  set(FREERTOS_PLUS_TCP_NETWORK_IF "ZYNQ" CACHE STRING "" FORCE)
endif()
set(FREERTOS_PLUS_TCP_TEST_CONFIGURATION "CUSTOM" CACHE STRING "" FORCE)
FetchContent_MakeAvailable(freertos_plus_tcp)
```

```

Related Issue
-----------
See #558 


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: CMakeLists.txt
Comment (aggarg) [inline CMakeLists.txt:14]: Is this a backward incompatible change? If so, is there a way to make it backward compatible?

File: CMakeLists.txt
Comment (aggarg) [inline CMakeLists.txt:200]: I am not Cmake expert but is it necessary to link `freertos_kernel` target when `freertos_kernel` links ``freertos_kernel_port` below.

File: CMakeLists.txt
Comment (aggarg) [inline CMakeLists.txt:225]: These do not seem applicable to all the compilers and therefore, should not be here.

File: portable/CMakeLists.txt
Comment (aggarg) [inline portable/CMakeLists.txt:1008]: Same as above.

File: CMakeLists.txt
Comment (archigup) [reply CMakeLists.txt:200]: It should since port uses kernel headers.

Comment (paulbartell) [reply CMakeLists.txt:200]: I don't think it makes sense to export both the port and kernel as separate STATIC lib targets. 

They're fully intertwined by the preprocessor, so it really only makes sense to export a static lib target for the combination of a config, port, and the kernel. 

If we're going to export targets for each piece separately, they should be INTERFACE lib targets rather than STATIC lib targets.

File: portable/CMakeLists.txt
Comment (paulbartell) [reply portable/CMakeLists.txt:1008]: I would rather leave this to the caller to handle. We don't want to unexpectedly mask any warnings.

File: CMakeLists.txt
Comment (phelter) [reply CMakeLists.txt:200]: First question: Yes it is necessary to link `freertos_kernel` target to `freertos_kernel_port` as `freertos_kernel_port` consumes the headers in `freertos_kernel`.  There is no issue with circular dependencies since it is linked as  `PRIVATE`.

Second question: Not quite sure what you refer to as exported.  It is necessary to separate the static libs since the `freertos_kernel_port` is select-able and it makes it easier if these are separate.  I was debating adding a CMakeLists.txt per `FREERTOS_PORT` type instead of having it monolithic in the port directory - since this limits the scope of a change to a particular port when a single port changes.

As for exporting each piece separately.  No - you can consume `freertos_kernel` in your `target_link_libraries` at the project level and you will also get both the `freertos_config` and the `freertos_kernel_port` linked too due to the PUBLIC link of those here: https://github.com/phelter/FreeRTOS-Kernel/blob/1822d8eb9aeb5aedb66cbf56e4a7fca58d898009/CMakeLists.txt#L241

A bit of a tutorial on `INTERFACE` lib targets - they `should` only be used for header only targets.  You could separate the `freertos_kernel` into an `INTERFACE` and `STATIC` lib but unless there is a really good reason - i.e. circular dependencies on targets, it is much easier to maintain as a single `target`

File: portable/CMakeLists.txt
Comment (phelter) [reply portable/CMakeLists.txt:1008]: The goal here was to in at least Clang and GNU compilers - enforce ALL warnings to be defined and then identify specifically the warnings that the `FreeRTOS-Kernel` group is comfortable masking.  All of these are currently masked in the current build since -Wall and -Wpedantic are currently not used.

The goal would be in further pull requests to:
- Fix any of the identified issues, OR 
- Identify the issues that you are willing to allow and move to the top level CMakeLists.txt:
  - eg: `-Wno-padded` and `-Wno-reserved-macro-identifier`
- Mask the ones out in header files that you are comfortable with by using local comments:
  - eg: 
```c
// NOLINTNEXTLINE disable-macro-expansion
```
- Note if you want them local to a particular port then I suggest making a CMakeLists.txt per port and then aliasing the `freertos_kernel_port` target to that particular port.

File: CMakeLists.txt
Comment (phelter) [reply CMakeLists.txt:14]: Yes - have added a warning if FREERTOS_CONFIG_FILE_DIRECTORY is used and if it is will add the appropriate info for freertos_config target if that has not been configured.

File: CMakeLists.txt
Comment (phelter) [reply CMakeLists.txt:247]: Disagree - when you have compiler checks for all the various compilers you are supporting you can add the other compilers in here similar to the way they are done here with the ID specified.  This is just Clang and GNU as those were the compilers I had available for my debugging on a linux machine.  This was here to be added to at a later date.

**Comment by phelter** [comment]
Thanks for your comments .  I've responded with all of them, and have allowed the FREERTOS_CONFIG_FILE_DIRECTORY to be backwards compatible - but with a warning to update when used.

If you like I can separate the ports each into separate CMakeLists.txt as well which would help simplify and limit the case statements in the port directory to just one (which one it selects).

File: portable/CMakeLists.txt
Comment (aggarg) [reply portable/CMakeLists.txt:1008]: I agree that which warnings to suppress should be decided by the application writer.

Comment (phelter) [reply portable/CMakeLists.txt:1008]: @aggarg and @paulbartell - How do you propose an application writer decide what the `FreeRTOS-Kernel` developer has deemed acceptable in the `freertos_kernel` code base? And how does the application writer decides what is an expected concern/lint warning and what is not?

As it stands now, compiling the `freertos_kernel` target has decided to ignore many warnings and linter checks through the omission of `-Wpedantic, -Wall, -Weverything` and using the default settings/warnings of the GNU and Clang compiler.
This PR is attempting to fix that issue and move to an opt-out approach instead of an opt-in approach so that the application developer isn't burdened with trying to figure out the assumptions of the `freertos_kernel` developer.

These suppressed warnings are a sub set of the warnings that were previously suppressed.

As mentioned above, the intent here is to inform what is currently necessary to suppress to allow compilation to occur with the `-Wall, -Wpedantic, etc.` enabled and without error.  At some point in the future, additional changes can be added to fix these issues and then remove the need to explicitly state that these are suppressed.

It allows further code changes to adhere to a stricter standard of coding.

If one continues to omit the `-Wall, -Weverything, -Wpedantic, etc.` the FreeRTOS developer is now dictating to the application developer that they are unable to compile the entire design with these settings for the compilers without manually adding these types of additions to each of the targets in the FreeRTOS-Kernel code base, thus causing a significant amount of work for every application developer as opposed to leaving these in and adding further ones for other compilers as they are added to the test suite of compilation (eg. CI/CD compile for ARMClang, CCS, IAR, etc.).  I only added the compilers that I have access to since the CI/CD environment does not actively test each and every compiler and port variation.

**Review by archigup — DISMISSED** [review]

File: portable/CMakeLists.txt
Comment (paulbartell) [reply portable/CMakeLists.txt]: @phelter An application writer can always apply the relevant compiler flags to a target using [target_compile_options](https://cmake.org/cmake/help/latest/command/target_compile_options.html).

This is particularly important since newer versions of gcc add new warnings (and suppression flags) all the time. So adding warning suppressions here is likely to cause build failures with older toolchains.

File: CMakeLists.txt
Comment (phelter) [reply CMakeLists.txt:222]: That is an unconventional way of using an INTERFACE library in CMake, but will still work.

The only thing that you'd need to change in this PR is:

From: `target_sources(freertos_kernel PRIVATE`
To: `target_sources(freertos_kernel PUBLIC`

All other changes are superficial and you do not require an INTERFACE type of target for this purpose.

What this does is it tells the builder to `copy a version of the code into the build directory of whatever will link to freertos_kernel and then compile out of the `build` directory of the target that linked to `freertos_kernel`.  This has many drawbacks.  The first being that you may want multiple other targets consuming `freertos_kernel` in the same application.  In this case `freertos_kernel` will be compiled multiple times for each target that links to it (and has the potential to be compiled differently with different compile options (not something you typically want) for two targets that then get consumed into a single application.  I HIGHLY RECOMMEND that you DO NOT take this approach.

If you would like to generate multiple binaries with different FreeRTOSConfig.h in the same cmake project then this is typically done with multiple compile options, eg:

```bash
cmake -GNinja -DSELECT_FREERTOS_APP=APP1 -B build_app1 .
cmake -GNinja -DSELECT_FREERTOS_APP=APP2 -B build_app2 .
```
In this case the application layer decides which version of `FreeRTOSConfig.h` is used for each APP1 and APP2 in its own `CMakeLists.txt` and the entire directory build_app1 is build with one implementation and the build_app2 directory could be a different implementation.  They both would consume the same code, just in different directories.  This is more common way of providing the feature of multiple binaries with different `FreeRTOSConfig.h` files and would be the suggested way to explain to integrators/application developers.

You could do it the other way as shown in your example of the `RPI Pico SDK` but then for 90% of the use cases where it is only one port that they require, you would be potentially burdening that single compilation with multiple additional compiles, and there is a larger chance an application developer could shoot themselves in the foot with different FreeRTOSConfig.h files for the same application.

This also leads down the path of then supporting a single cmake project wanting two FreeRTOS-Kernels with two different ports - eg. Little-Big ARM based SoC's.  That situation can be handled with the 'multiple compile options` method mentioned above too.

File: portable/CMakeLists.txt
Comment (phelter) [reply portable/CMakeLists.txt]: Thanks @paulbartell - I've removed them and will unfortunately have to add them into every application project now consuming freertos kernel.

How does FreeRTOS-Kernel check and identify the supported compilers and versions will in-fact compile?

File: CMakeLists.txt
Comment (phelter) [reply CMakeLists.txt]: They are removed.

**Comment by phelter** [comment]
Note - I have a follow-on PR for fixing the compiler warnings here: https://github.com/phelter/FreeRTOS-Kernel/pull/1  will change target to master once this PR is completed.

**Comment by phelter** [comment]
@paulbartell @aggarg - Any chance we can move the needle on this PR?

I've performed all of the changes you've requested. And am waiting for this one to be merged to start the next one.

Thanks

**Comment by phelter** [comment]
@paulbartell @aggarg - Ping again.

Need a maintainer to approve running the workflows for this PR and would appreciate some time to identify any further issues that need to be resolved.

Thanks

**Review by AniruddhaKanhere — DISMISSED** [review]

**Review by archigup — DISMISSED** [review]

File: CMakeLists.txt
Comment (paulbartell) [inline CMakeLists.txt:36]: I would much prefer that we leave this up to the caller to handle and just deprecate FREERTOS_CONFIG_FILE_DIRECTORY.

Then the caller can add their relevant FreeRTOSConfig.h file and directory to the target in whichever way they see fit.

File: CMakeLists.txt
Comment (paulbartell) [inline CMakeLists.txt:208]: We often build GCC_POSIX with Mingw, so I would prefer to only include this if using MSVC.

File: CMakeLists.txt
Comment (paulbartell) [reply CMakeLists.txt:228]: Multiple build directories are sometimes a reasonable option, but not the only option. I don't think it makes sense to enforce this on users, especially when it will further complicate adding CI checks for each port and their multitude of different configurations.

If we're set on exporting a freertos_kernel static library target, we should just include the selected port port with it rather than creating two separate targets and implying that the two are not circularly dependent on each other.

File: CMakeLists.txt
Comment (paulbartell) [inline CMakeLists.txt:255]: IMO, this should be left to the user.

**Review by paulbartell — CHANGES_REQUESTED** [review]
IMO, defining and linking a target similar to freertos_config should be left to the user.

File: CMakeLists.txt
Comment (phelter) [reply CMakeLists.txt:208]: This code is only called if FREERTOS_PORT is undefined in the original cmake build step.

If you would like to use GCC_POSIX with Mingw - then that can be done by: `cmake -DFREERTOS_PORT=GCC_POSIX ....` when calling it from MINGW prompt.  And explicitly defining the `FREERTOS_PORT`.

When using `MINGW` as the platform, the most likely candidate is `MSVC_MINGW` - so that is why that was selected when it is not explicitly defined.

File: CMakeLists.txt
Comment (phelter) [reply CMakeLists.txt:228]: @paulbartell - If FreeRTOS is wanting to support multiple Ports, then shouldn't the CI checks confirm each port can be compiled - with some minimal checks - eg: `Default, AllEnabled, AllDisabled` - for the portConfig info?  And when a user decides to have their own special `config` and a bug related to the kernel is found in their `freertos_config` configuration - it is an easy fix to confirm the bug by providing that user's config.h files to re-create.

At this moment nothing is enforced on the user/consumer -  The changes here are following the original CMake that was provided.  The only difference is that the user now specify their one and only one `freertos_config` target instead of defining one and only one `FREERTOS_CONFIG_FILE_DIRECTORY` that was specified. One cannot have multiple values of a CMAKE variable for the same build directory.

As mentioned - I'm not sure what you are referring to as `exporting` as this is expected to be compiled with the user's code (as based on the `freertos_config`) so there is no notion of `exporting` - which is typically a common lib that is provided compiled without configuration.  As mentioned above, the `freertos_kernel` target is an aggregate of the `freertos_kernel_port` (the one selected), the `freertos_config` - provided by the user, and the `freertos_kernel` code.  - This has not changed from the original definition of CMake files as they already had an aggregate dependency as mentioned.

Within CMake there is no means of hiding a target from the user, so the `freertos_kernel_port` being defined independently will be visible to the user, and because of the dependencies will compile both `freertos_kernel` and `freertos_kernel_port`.

If you wanted to remove what you are referring to as `circular-dependencies` one could have:
- create a `freertos_kernel_port_api` which would only be the header files that are consumed by the `freertos_kernel_port_x`
- include `freertos_kernel_port_api` as part of both `freertos_kernel_port_x` and `freertos_kernel`
- Deal with the confusion of having yet another target.
Typically the `api` is there if you intend on supporting more than one implementation - you define `api` so that that is the base for any dependencies.

But this is all out of the scope of what I was trying to achieve.

1. Support for a target instead of a single PATH to the config files.
2. Support for a specialized port that isn't defined as one of the defaults in the FreeRTOS-Kernel
- Where the second one would mean writing your own CMake if you wanted to do so instead of relying on what is already there.

File: CMakeLists.txt
Comment (phelter) [reply CMakeLists.txt:255]: @paulbartell  - Agree - Yes - it is left to the user - just as it was left to the user to define `FREERTOS_CONFIG_FILE_DIRECTORY`.

**Comment by phelter** [comment]
> IMO, defining and linking a target similar to freertos_config should be left to the user.

@paulbartell  - You've requested a change.  I am unclear as to what that change is that will get approval for this PR.

Are you asking NOT to link `freertos_config` to `freertos_kernel` ?

If this is the case, can you please provide a proposal of how to provide a standardized way a `user` is to define a target similar to the `freertos_config` and provide more details as to what they must do to integrate their config with `freertos_kernel`.  

I am unclear as to what must be done next to move this PR along.

Thanks

File: CMakeLists.txt
Comment (paulbartell) [reply CMakeLists.txt:228]: @phelter : I understand your position and I'm very glad to have you contributing. The CMakeLists.txt is not perfect, but we're making good progress.

One reuqest, Could you make the freertos_config target optional and link to it if and only if it exists? With that change, I can approve this PR.

Comment (phelter) [reply CMakeLists.txt:220]: @paulbartell - Those changes are now in, and I also identified the deprecated `FREERTOS_CONFIG_FILE_DIRECTORY` portions of the cmake, and removed the auto generation of the `freertos_config` target if a user had `FREERTOS_CONFIG_FILE_DIRECTORY` defined prior but no `freertos_config`.

**Review by paulbartell — DISMISSED** [review]

**Comment by phelter** [comment]
@AniruddhaKanhere , @aggarg  or @archigup - would it be possible for one of you to approve this PR now that all of the issues/changes requested by @paulbartell are now resolved.

Thanks.

**Comment by phelter** [comment]
@AniruddhaKanhere , @aggarg or @archigup - Ping again.  Would appreciate a review of the changes requested by @paulbartell

**Comment by phelter** [comment]
@AniruddhaKanhere , @aggarg or @archigup - Ping once again.  Just so you know this has been opened closing in on almost 2 months now.

**Review by AniruddhaKanhere — APPROVED** [review]

**Comment by AniruddhaKanhere** [comment]
@phelter, apologies for the delay, I was out of office and was travelling.
I approve of your change!

Thank you for your patience and your efforts in creating this PR.

Thanks,
Aniruddha

**Review by paulbartell — APPROVED** [review]

**Pull request merged**
